lib/internal/modules/esm/translators.js

```diff
@@ -276,11 +285,37 @@ function createCJSModuleWrap(url, source, isMain, loadCJS = loadCJSModule) {
     setOwnProperty(process, 'mainModule', module);
   }
 
-  return new ModuleWrap(url, undefined, namesWithDefault, function() {
-    debug(`Loading CJSModule ${url}`);
-
-    if (!module.loaded) {
+  let moduleLoadingError;
+  debug(`Loading CJSModule ${url}`);
+  if (!module.loaded) {
+    try {
       loadCJS(module, source, url, filename);
+    } catch (err) {
+      /**
+       * If this module contained ESM syntax, we will retry loading it as an ES module if the following conditions are met:
+       * - For file input:
+       *   - It is in a package scope with no defined `type`: neither `module` nor `commonjs`.
+       *   - It does not have an explicit .cjs or .mjs extension.
+       * - For string input:
+       *   - `--input-type` is unset.
+       *   - `--experimental-default-type` is unset.
+       */
+      if (detectModule && isModuleSyntaxError(err) && (
+          (isFileURL && getPackageType(url) === 'none' && !filename.endsWith('.mjs') && !filename.endsWith('.cjs')) ||
+          (!isFileURL && inputTypeFlagInputValue === '' && defaultTypeFlagInputValue === '')
+        )) {
+        // We need to catch this error here, rather than during the ModuleWrap callback, in order to bubble this up so that we
+        // can retry loading the module as an ES module.
+        throw err;
+      }
+      moduleLoadingError = err;
+    }
+  }
+
+  return new ModuleWrap(url, undefined, namesWithDefault, function() {
+    // The non-detection flow expects module loading errors to be thrown here.
+    if (moduleLoadingError) {
+      throw moduleLoadingError;
```

Comment:
will catching and rethrowing this error trigger the correct callsite locator magic node has?